### PR TITLE
Add VGMdb support

### PR DESCRIPTION
**Pull request by Eldarock**
Add vgmdb support for matching and fetchart.
HIGHLY ALPHA, just wanted to do a PR because i want your feedbacks and don't know where exactly to contact you.
Let me know what is wrong (it's like a "night coding session" work), i'm already using it on my massive VGM library.

All comments are welcomed, just started python 2 days ago.

Credits to early vgmdb plugin : 
https://github.com/whtsky/VGMdb-BeetsPlug
https://github.com/SomeTiramisu/VGMdb-BeetsPlug

### CODE REVIEW

File: beets/autotag/hooks.py [hunk 1]
```diff
@@ -78,7 +78,8 @@ def __init__(self, tracks, album=None, album_id=None, artist=None,
                  original_year=None, original_month=None,
                  original_day=None, data_source=None, data_url=None,
                  discogs_albumid=None, discogs_labelid=None,
-                 discogs_artistid=None, **kwargs):
+                 discogs_artistid=None, vgmdb_album_id=None,
+                 vgmdb_artist_id=None, vgmdb_game=None, **kwargs):
```
Comment (sampsyo) [inline beets/autotag/hooks.py:82]: There should be no need to add any new fields here (nor to touch `hooks.py` at all); the new data can just go in `kwargs`.

File: beets/library.py [hunk 2]
```diff
@@ -468,6 +468,9 @@ class Item(LibModel):
         'discogs_albumid': types.INTEGER,
         'discogs_artistid': types.INTEGER,
         'discogs_labelid': types.INTEGER,
+        'vgmdb_album_id': types.INTEGER,
+        'vgmdb_artist_id': types.INTEGER,
+        'vgmdb_game': types.STRING,
```
Comment (sampsyo) [inline beets/library.py:473]: There's also no need to add fixed attributes for the new fields. Let's just keep them as flexible attributes (meaning that the plugin, not `library.py`, will provide the types).

File: beetsplug/fetchart.py [hunk 3]
```diff
@@ -401,6 +401,23 @@ def get(self, album, plugin, paths):
...
 
 
+class VGMdb(RemoteArtSource):
+    NAME = "VGMdb"
+    URL = 'https://vgmdb.info/album/%s?format=json'
+    INDICES = (1)
+
+    def get(self, album, plugin, paths):
+        print("SLT")
+        print(album.vgmdb_album_id)
```
Comment (sampsyo) [inline beetsplug/fetchart.py:411]: Looks like some stray debugging output.

File: beetsplug/fetchart.py [hunk 4]
```diff
@@ -678,7 +695,7 @@ def get(self, album, plugin, paths):
         # This may be removed once the DBPedia issue is resolved, see:
         # https://github.com/dbpedia/extraction-framework/issues/396
         if ' .' in cover_filename and \
-           '.' not in cover_filename.split(' .')[-1]:
+                '.' not in cover_filename.split(' .')[-1]:
```
Comment (sampsyo) [inline beetsplug/fetchart.py:698]: These whitespace changes look unrelated; let's leave them out.

File: docs/changelog.rst [hunk 5]
```diff
@@ -30,6 +30,9 @@ Major new features:
 * :doc:`/plugins/albumtypes`: An accompanying plugin for formatting
   ``albumtypes``. Thanks to :user:`edgars-supe`.
 
+* Add new VGMdb source for match and fetchart
+  :user:`eldarock`.
```
Comment (sampsyo) [inline docs/changelog.rst:34]: Seems like it should probably be two bullet points: one for the new plugin and one for the new feature in `fetchart`.

File: beetsplug/vgmdb.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
+# update to python3
+# use lang priority everywhere
+# add prefer_original setting
```
Comment (sampsyo) [inline beetsplug/vgmdb.py:3]: Please use the standard copyright header.

File: beetsplug/vgmdb.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+log = logging.getLogger('beets')
+
+
+class VGMdbPlugin(BeetsPlugin):
+
+    def __init__(self):
+        super(VGMdbPlugin, self).__init__()
+        self.config.add({
+            'source_weight': 1.0,
+            'lang-priority': 'en, ja-latn, ja',
```
Comment (sampsyo) [inline beetsplug/vgmdb.py:24]: Any particular reason this is a string and not a proper list?

File: beetsplug/vgmdb.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+            'prefer_original': True  # Only for Attack on Titan soundtracks
```
Comment (sampsyo) [inline beetsplug/vgmdb.py:25]: That's a somewhat suspicious comment—maybe this needs documentation?

File: beetsplug/vgmdb.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        })
+        log.debug('Querying VGMdb')
+        self.source_weight = self.config['source_weight'].as_number()
+        self.lang = self.config['lang-priority'].get().replace(' ', '').split(",")
```
Comment (sampsyo) [inline beetsplug/vgmdb.py:29]: Could use `as_str_seq()` if this option were a proper list.

File: beetsplug/vgmdb.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        """
+        if va_likely:
+            query = album
+        else:
+            query = '%s %s' % (artist, album)
+        try:
+            ret = self.get_albums(query, va_likely)
+            for x in ret:
+                msg = 'Found album: %s ; id: %s ; artist: %s' % (x.album, x.album_id, x.artist)
+                log.debug(msg)
```
Comment (sampsyo) [inline beetsplug/vgmdb.py:81]: Seems like this could be one line?

File: beetsplug/vgmdb.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        """
+        log.debug('Querying VGMdb for release %s' % str(album_id))
+
+        # Get from VGMdb
+        r = requests.get('http://vgmdb.info/album/%s?format=json' % str(album_id))
+
+        # Decode Response's content
+        try:
+            item = r.json()
+        except:
```
Comment (sampsyo) [inline beetsplug/vgmdb.py:99]: Plain `except:` blocks are dangerous—perhaps handle a `JSONDecodeError` or whatever it's called?

File: beetsplug/vgmdb.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+        # Break up and get search results
+        for item in items["results"]["albums"]:
+            album_id = str(self.decod(item["link"][6:]))
+            albums.append(self.album_for_id(album_id))
+            if len(albums) >= 5:
+                break
+        log.debug('get_albums Querying VGMdb for release %s' % str(query))
+        return albums
+
+    def decod(self, val, codec='utf8'):
```
Comment (sampsyo) [inline beetsplug/vgmdb.py:137]: Any particular reason not to use the full word, `decode`?

File: beetsplug/vgmdb.py [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+        """Ensure that all string are coded to Unicode.
+        """
+        if isinstance(val, str):
+            b = val.encode()
+            return b.decode(codec, 'ignore')
```
Comment (sampsyo) [inline beetsplug/vgmdb.py:142]: I'm a little unclear about the purpose here. If it's already a `str`, why bother encoding it and decoding again? In what cases is this invoked on something that is *not* a `str`, and why would it be left unchanged?

**Review by sampsyo — COMMENTED** [review]
Looks good overall! I have a few suggestions within. Another step will be to add documentation for the new plugin and to update the documentation for the `fetchart` plugin.

File: beetsplug/fetchart.py [hunk 14]
```diff
@@ -401,6 +401,23 @@ def get(self, album, plugin, paths):
...
+    URL = 'https://vgmdb.info/album/%s?format=json'
+    INDICES = (1)
+
+    def get(self, album, plugin, paths):
+        print("SLT")
+        print(album.vgmdb_album_id)
+        if album.vgmdb_album_id:
+            try:
+                resp = self.request(self.URL % album.vgmdb_album_id)
+                url = resp.json()['picture_full']
```
Comment (wisp3rwind) [inline beetsplug/fetchart.py:415]: I suspect that `KeyError`s from this line should also be caught, in case vgmdb has not artwork for an album? Maybe, you can find an example of an album without artwork and see how this is represented in the JSON?

File: beetsplug/fetchart.py [hunk 15]
```diff
@@ -909,7 +929,7 @@ def __init__(self):
             'cautious': False,
             'cover_names': ['cover', 'front', 'art', 'album', 'folder'],
             'sources': ['filesystem',
-                        'coverart', 'itunes', 'amazon', 'albumart'],
+                        'coverart', 'itunes', 'amazon', 'vgmdb', 'albumart'],
```
Comment (wisp3rwind) [inline beetsplug/fetchart.py:932]: I'd argue for not changing the defaults here, since VGMDB seems like a relatively niche service, which many users won't need.

**Review by wisp3rwind — CHANGES_REQUESTED** [review]
I've added a few small comments on the fetchart additions, but didn't look at the autotagger.
